browser/notifications/notification_platform_bridge_brave_custom_notification.cc

```diff
@@ -0,0 +1,178 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/notifications/notification_platform_bridge_brave_custom_notification.h"
+
+#include <memory>
+#include <string>
+
+#include "base/bind.h"
+#include "base/callback.h"
+#include "base/memory/scoped_refptr.h"
+#include "base/run_loop.h"
+#include "base/task/post_task.h"
+#include "brave/components/brave_ads/browser/ads_notification_handler.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_ads/browser/ads_service_factory.h"
+#include "brave/components/brave_ads/browser/ads_service_impl.h"
+#include "brave/ui/brave_custom_notification/message_popup_view.h"
+#include "brave/ui/brave_custom_notification/public/cpp/notification.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/notifications/notification_ui_manager.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+
+#if defined(OS_ANDROID)
+#include "base/android/jni_android.h"
+#include "base/android/jni_array.h"
+#include "base/android/jni_string.h"
+#include "brave/browser/brave_ads/android/jni_headers/BraveAdsCustomNotificationDialog_jni.h"
+using base::android::AttachCurrentThread;
+using base::android::ConvertJavaStringToUTF8;
+using base::android::ConvertJavaStringToUTF16;
+using base::android::ConvertUTF16ToJavaString;
+using base::android::ConvertUTF8ToJavaString;
+using base::android::JavaParamRef;
+using base::android::ScopedJavaLocalRef;
+#endif
+
+namespace {
+
+class PassThroughDelegate
+    : public brave_custom_notification::NotificationDelegate {
+ public:
+  PassThroughDelegate(
+      Profile* profile,
+      const brave_custom_notification::Notification& notification)
+      : profile_(profile),
+        notification_(notification) {}
+
+  void Close(bool by_user) override {
+    std::unique_ptr<brave_ads::AdsNotificationHandler> handler =
+        std::make_unique<brave_ads::AdsNotificationHandler>(
+            static_cast<content::BrowserContext*>(profile_));
+    handler->SetAdsService(
+      static_cast<brave_ads::AdsServiceImpl*>(
+        brave_ads::AdsServiceFactory::GetForProfile(profile_)));
```

Comment:
We are calling `GetForProfile` here and `GetImplForProfile` in other places, please see my previous comment on line 57